python/pyspark/sql/tests.py

```diff
@@ -4509,23 +4523,32 @@ def weighted_mean(v, w):
         return weighted_mean
 
     def test_manual(self):
+        from pyspark.sql.functions import pandas_udf, array
+
         df = self.data
         sum_udf = self.pandas_agg_sum_udf
         mean_udf = self.pandas_agg_mean_udf
-
-        result1 = df.groupby('id').agg(sum_udf(df.v), mean_udf(df.v)).sort('id')
+        mean_arr_udf = pandas_udf(
+            self.pandas_agg_mean_udf.func,
```

Comment:
For arrays, can we add tests for:
* Test type coercion, e.g., specified type is `array<double>` and returned array is `[0, 1, 2]`?
* Test exception: function returns array of different types like `[0, "hello"]`

Reply:
Btw, we can have this to be a follow up and I can do it too

Reply:
If you meant to type coercion (did I understand correctly?), I already tested. Seems not working properly. Similar thing was discussed in https://github.com/apache/spark/pull/20163#issuecomment-356231655 (thanks @ueshin).

Will reread the comments when I am more awake tomorrow ...

Reply:
I think with Pandas UDFs, certain type coercion is supported, e.g., when user specify "double type" and returns a `pd.Series` of `int`, it will automatically cast it to `pd.Series` of double. This behavior is different from regular Python UDF, which will return null in this case. Most of the type coercion is done by pyarrow. (Btw, I think type coercion in Pandas UDFs is an huge improvement over Python UDF because that's one of the biggest frustration our PySpark users have...)

Reply:
Btw, if type coercion is not working with array type, I think it's still fine to allow using array type and fix type coercion separately.

Reply:
Hm .. let's do it separately for type coercion stuff in another issue. I think we need another iteration for it.